### PR TITLE
extensions checks: Add OCSP must staple check.

### PR DESCRIPTION
**Pull request by cpu**
This commit adds a check for the OCSP Must Staple x509v3 feature
extension defined in RFC 7633. Prior to this commit running `certlint`
on a certificate with this extension produced output like:

```
Bad Certificates Detected: {
  'xxxx': {
    'valid': False,
    'problems': [
      'Certificate contains unknown extension (1.3.6.1.5.5.7.1.24)'
    ]
  }
}
```

The OCSP Must Staple check presently enforces that:
1) The `certdata.Data.Type` is not OCSP. The RFC only defines this
   extension for end entity certificates, CSRs, and issuing
   certificates, not OCSP signing certificates.
2) The `ex.Critical` flag is set to false. The RFC forbids setting this
   extension to critical.

After applying this commit the above Bad Certificate Detected problem is
resolved.

_*Note to reviewers:* I did not find any unit tests for the existing 
`certlint/extensions/` checks so I haven't tried to write one for this 
addition. I did verify using known good certificates both with/without 
the extension. If there is an example unit test to refer to I can attempt
to add one for the OCSP Must Staple check._

### CODE REVIEW

File: checks/extensions/ocspmuststaple/ocspmuststaple.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	checks.RegisterExtensionCheck(checkName, extensionOid, nil, Check)
+}
+
+// Check performs a strict verification on the extension according to the standard(s)
+func Check(ex pkix.Extension, d *certdata.Data) *errors.Errors {
+	var e = errors.New(nil)
+
+	// RFC 7633 only defines this extension for PKIX end-entity certificates,
+	// certificate signing requests, and certificate signing certificates (CAs)
+	if d.Type == "OCSP" {
```
Comment (vanbroup) [inline checks/extensions/ocspmuststaple/ocspmuststaple.go]: You might want to check IsCa and or the key usage here, the current check is is only covering an edge case.

**Comment by cpu** [comment]
> As this is extension is not supported by Go we should be checking more than the critical flag. It should be not to hard to test test the extension value against an expected byte array or to unmarshal the value.

@vanbroup This is sensible feedback & would make a much better check. My preference would be to merge this as-is and file a follow-up issue to enhance the implementation. I may not have time to return to this PR for a few days and I'd like to unblock users with OCSP Must Staple certificates. Are you open to that?

For reference I believe this check & the level of scrutiny it applies is fairly consistent with the existing codebase. I based this off of [`ocspnocheck.go`](https://github.com/globalsign/certlint/blob/master/checks/extensions/ocspnocheck/ocspnocheck.go) which seems like it would need the same review feedback you gave this work. It does not check the `IsCa` field of the cert, or its key usage: 
https://github.com/globalsign/certlint/blob/c11b63719ad75636b9a29854164c2bdf5c113b0b/checks/extensions/ocspnocheck/ocspnocheck.go#L24-L26
Similarly, though the extension is not supported by Go there is no explicit check of the ASN.1. Many other checks appear to apply the same surface level evaluation I did in this PR (e.g. [adobetimestamp.go](https://github.com/globalsign/certlint/blob/master/checks/extensions/adobetimestamp/adobetimestamp.go), [pdfrevocation.go](https://github.com/globalsign/certlint/blob/master/checks/extensions/pdfrevocation/pdfrevocation.go), [smimecapabilities.go](https://github.com/globalsign/certlint/blob/master/checks/extensions/smimecapabilities/smimecapabilities.go)). It would be helpful to external contributors if there were more guidelines available for writing checks & consistency throughout the existing checks.

Thanks!

File: checks/extensions/ocspmuststaple/ocspmuststaple.go (hunk 1 above)
Comment (vanbroup) [reply checks/extensions/ocspmuststaple/ocspmuststaple.go]: The check `d.Type == "OCSP"` doesn't check for end-entity certificates, it would only apply for an OCSP signing certificate. I think you are confused with the OCSPNoCheck extension who MUST only be applied on OCSP signing certificates.

**Comment by vanbroup** [comment]
> For reference I believe this check & the level of scrutiny it applies is fairly consistent with the existing codebase. I based this off of ocspnocheck.go which seems like it would need the same review feedback you gave this work. It does not check the IsCa field of the cert, or its key usage:

As added to the review comment, OCSPNoCheck applies only in OCSP signing certificates and not to any end-entity certificates. The type OCSP is based on the extended key usage `x509.ExtKeyUsageOCSPSigning`. If you just check for `IsCA: false` than you directly cover all end-entity certificates.

> Similarly, though the extension is not supported by Go there is no explicit check of the ASN.1. Many other checks appear to apply the same surface level evaluation I did in this PR (e.g. adobetimestamp.go, pdfrevocation.go, smimecapabilities.go). It would be helpful to external contributors if there were more guidelines available for writing checks & consistency throughout the existing checks.

I agree that is not completely fair, I think most recent changes have all come with test coverage as I think it's important to work towards a 100% coverage for this package. The test case required for this extension is significantly simple that I feel it should not be much work to add.

File: checks/extensions/ocspmuststaple/ocspmuststaple.go (hunk 1 above)
Comment (cpu) [reply checks/extensions/ocspmuststaple/ocspmuststaple.go]: > The check d.Type == "OCSP" doesn't check for end-entity certificates, it would only apply for an OCSP signing certificate.

Yup. I wasn't trying to explicitly check for end-entity certificates with this clause. That's why the comment says "end-entity certificates, certificate signing requests, and certificate signing certificates".  I was trying to find certificates that were not end-entity, CSRs, or CA certificates. In my scan of `certdata/types.go` it seemed like `OCSP` was the only type that didn't meet those criteria so I checked for it explicitly.

I'm going to dump this check entirely. The `certdata/type.go` package and its types seem too coarse to be especially useful here and your suggestion to check the key usage explicitly is probably better.

**Comment by cpu** [comment]
>  If you just check for IsCA: false than you directly cover all end-entity certificates.

Sounds good :+1: I'll do that instead.

>  I think most recent changes have all come with test coverage as I think it's important to work towards a 100% coverage for this package. The test case required for this extension is significantly simple that I feel it should not be much work to add.

Ok, I can work with that. I'll close this PR and re-open it when I have time to work on adding a test case.

Thanks,
